### PR TITLE
Feature/175 git submodule for openzeppelin contracts

### PR DESCRIPTION
**Pull request by apetrovskiy**
1. work via a submodule - done
2. openzeppelin-contracts working via hardhat - done (by theirselves)
3. support for neonlabs in hardhat.config.js - done
image
fixed
4. where to save - now yet done
- hardhat.config.js
- truffle config
- migration scripts (???)

image

### CODE REVIEW

File: hardhat.config.js [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+process.env.ADDRESS_TO = account02.address;
+
+const privateKeys = [process.env.PRIVATE_KEY,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey];
```
Comment (mich-master) [inline hardhat.config.js]: Up to 10 accounts required.

File: hardhat.config.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    },
+  },
+  defaultNetwork: "neonlabs",
+  networks: {
+    hardhat: {
+      blockGasLimit: 10000000,
+      allowUnlimitedContractSize: !withOptimizations,
+    },
+    neonlabs: {
+      url: process.env.PROXY_URL,
```
Comment (mich-master) [inline hardhat.config.js:97]: "accounts" field with privateKeys array required to avoid "get_accounts" error.
Other fields required optionally.

**Review by mich-master — COMMENTED** [review]
Ок

File: hardhat.config.js (hunk 1 above)
Comment (rozhkovdmitrii) [reply hardhat.config.js]:
```js
const accounts = Array.from(Array(10), (_, x) => web3.eth.accounts.create());
const privateKeys = accs.map((account) => account.privateKey);
```
Wouldn't it be better?

File: hardhat.config.js [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+    currency: 'USD',
+    outputFile: argv.ci ? 'gas-report.txt' : undefined,
+    coinmarketcap: argv.coinmarketcap,
+  },
+  mocha: {
+    timeout: 1000000000,
+    reporter: 'mocha-multi-reporters',
+    reporterOption: {
+      configFile: '../reporterConfig.json',
+      // url: process.env.PROXY_URL,
```
Comment (rozhkovdmitrii) [inline hardhat.config.js]: If drop this comment it will be clear code

File: hardhat.config.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey];
+
+// const faucetUrl = process.env.PROXY_URL.replace("/solana", "/request_erc20_tokens");
```
Comment (rozhkovdmitrii) [inline hardhat.config.js]: I'm not sure it's really need here

File: hardhat.config.js (hunk 1 above)
Comment (mich-master) [reply hardhat.config.js]: Looks worse for understanding without any advantage

File: hardhat.config.js [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+
+for (const f of fs.readdirSync(path.join(__dirname, 'hardhat'))) {
+  require(path.join(__dirname, 'hardhat', f));
+}
+
+const withOptimizations = argv.enableGasReport || argv.compileMode === 'production';
+
+const web3 = new Web3(new Web3.providers.HttpProvider(process.env.PROXY_URL));
+const account01 = web3.eth.accounts.create();
+process.env.ADDRESS_FROM = account01.address;
```
Comment (rozhkovdmitrii) [inline hardhat.config.js:62]: Is it really needed to be declared as part of environment?
If not, it should be dropped, then you can get required authorities from accounts array

File: hardhat.config.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+process.env.PRIVATE_KEY = account01.privateKey;
+const account02 = web3.eth.accounts.create();
+process.env.ADDRESS_TO = account02.address;
```
Comment (rozhkovdmitrii) [inline hardhat.config.js:65]: Is it really needed to be declared as part of environment?

File: hardhat.config.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey,
+                     web3.eth.accounts.create().privateKey];
+
+// const faucetUrl = process.env.PROXY_URL.replace("/solana", "/request_erc20_tokens");
+const faucetUrl = process.env.FAUCET_URL
+console.log(faucetUrl);
```
Comment (rozhkovdmitrii) [inline hardhat.config.js]: `console.log("FaucetUrl: " + facuetUrl)`

File: hardhat.config.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+const requestFaucet = (address, amount) => axios.post(faucetUrl, { wallet: address, amount: amount })
```
Comment (rozhkovdmitrii) [inline hardhat.config.js]: Can't find where this functor used in this script

File: hardhat.config.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    reporter: 'mocha-multi-reporters',
+    reporterOption: {
+      configFile: '../reporterConfig.json',
+      // url: process.env.PROXY_URL,
+    },
+    diff: true,
+  },
+};
+
+if (argv.coverage) {
```
Comment (rozhkovdmitrii) [inline hardhat.config.js:123]: Shouldn't it be on the top of script?

File: hardhat.config.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+// - COMPILE_VERSION:   compiler version (default: 0.8.3)
+// - COINMARKETCAP:     coinmarkercat api key for USD value in gas report
+
+require("dotenv").config({ path: "../.env" });
+const Web3 = require("web3");
+const axios = require('axios').default;
+
+const fs = require('fs');
+const path = require('path');
+const argv = require('yargs/yargs')()
```
Comment (rozhkovdmitrii) [inline hardhat.config.js:14]: It's better to initialize argv in three steps, isn't it?
1. `const yargs = require('yargs/yargs`)
2. `const cfg = yargs(...)`
3. `const argv = cfg.argv`

**Review by rozhkovdmitrii — CHANGES_REQUESTED** [review]

File: hardhat.config.js (hunk 1 above)
Comment (rozhkovdmitrii) [reply hardhat.config.js]: > Looks worse for understanding without any advantage

Looks like a copypaste without a reason )

File: hardhat.config.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+  networks: {
+    hardhat: {
+      blockGasLimit: 10000000,
+      allowUnlimitedContractSize: !withOptimizations,
+    },
+    neonlabs: {
+      url: process.env.PROXY_URL,
+      accounts: privateKeys,
+      from: process.env.ADDRESS_FROM,
+      network_id: process.env.NETWORK_ID,
```
Comment (rozhkovdmitrii) [inline hardhat.config.js:101]: Can't find `network_id` in  there https://hardhat.org/config/#json-rpc-based-networks as well as `allowUnlimitedContractSize`

File: hardhat.config.js (hunk 1 above)
Comment (apetrovskiy) [reply hardhat.config.js]: Fixed

File: hardhat.config.js (hunk 3 above)
Comment (apetrovskiy) [reply hardhat.config.js]: fixed

File: hardhat.config.js (hunk 4 above)
Comment (apetrovskiy) [reply hardhat.config.js]: fixed

File: hardhat.config.js (hunk 5 above)
Comment (apetrovskiy) [reply hardhat.config.js:62]: fixed

File: hardhat.config.js (hunk 6 above)
Comment (apetrovskiy) [reply hardhat.config.js:65]: fixed

File: hardhat.config.js (hunk 7 above)
Comment (apetrovskiy) [reply hardhat.config.js]: fixed

File: hardhat.config.js (hunk 9 above)
Comment (apetrovskiy) [reply hardhat.config.js:123]: It's shipped within the original hardhat.config.js

File: hardhat.config.js (hunk 11 above)
Comment (apetrovskiy) [reply hardhat.config.js:101]: fixed

File: hardhat.config.js (hunk 10 above)
Comment (apetrovskiy) [reply hardhat.config.js:14]: I don't have time to test the whole expression whether it's broken after my changes or not
image

File: hardhat.config.js (hunk 2 above)
Comment (apetrovskiy) [reply hardhat.config.js:97]: fixed

File: hardhat.config.js (hunk 8 above)
Comment (apetrovskiy) [reply hardhat.config.js]: removed

**Review by rozhkovdmitrii — APPROVED** [review]

**Pull request merged**